### PR TITLE
Pull Request to integrate read only HEIF support in Exiv2 using libheif

### PR DESCRIPTION
**Pull request by cgilles**
See my progress described in bug #318 

Best

Gilles Caulier

### CODE REVIEW

**Comment by phako** [comment]
Why are you including libheif in this?

**Comment by cgilles** [comment]
See my comments in #318...

File: src/heifimage_int.hpp
Comment (piponazo) [inline src/heifimage_int.hpp]: Other image headers such as `jp2image.hpp` include the header `image.hpp` where the class `Image` is defined. When opening this file in an IDE, I get lot of errors from the clang static analyser regarding files not found.

Furthermore, It looks like this file should be named `heifimage.hpp` without the `_int`suffix , and be moved to the folder `include/exiv2` in order to be deployed together with the other supported file format headers.

File: src/heifimage.cpp
Comment (piponazo) [inline src/heifimage.cpp]: [suggestion] This function has almost 200 LoC. Would it be possible to extract some blocks of code into separated private methods or functions?

File: src/libheif/LibHeifRules.cmake
Comment (piponazo) [inline src/libheif/LibHeifRules.cmake]: As mentioned previously I am against adding libheif directly in our repository. Another argument for that is that this cmake file is adding definitions and include directories globally to the project (instead of doing so to a specific target or source files).

File: src/heifimage_int.hpp
Comment (piponazo) [inline src/heifimage_int.hpp]: :+1:

**Review by piponazo — CHANGES_REQUESTED** [review]
Thanks for the contribution!

In general it looks good to me, but as I mentioned before, I do not like the idea of adding source code to an external library inside our repository. The main reason for that is that we could not upgrade easily to more recent versions of the library by just requiring a more recent version of it.

I will work during the next days in a **conan** recipe for this project so that we can consume the library easily. I expect to come back with some feedback, at the latest, this Sunday.

**Review by D4N — CHANGES_REQUESTED** [review]
I am definitely against bundling yet another library into exiv2's code base: we already bundle the xmpsdk and an INI parser. Also, if we bundle libheif, then we'll make packaging exiv2 extremely painful, as libheif is considered a no-go for many Linux distributions (you won't find it in Fedora for legal reasons and they got a dedicated legal team...). By bundling it, packagers will have to manually remove libheif, which is not a nice thing to do from our side.

Other general comments:
- commit f46967ddca0837794449c072751fe1d0ea1abb14 is already on master, no need to include it here
- please drop the merge commits and rebase+squash the `polish` commits
- please restore the original history from @1div0's branch, so that his edits are visible
- there is **a lot** of changes in `po/nl.po` that look completely unrelated to this PR, please consider to resubmit them in a different PR or drop them
- no tests at all

**Comment by piponazo** [comment]
The first versions of the conan package are available here:
https://bintray.com/beta/#/piponazo/piponazo/libheif:piponazo?tab=overview

Please, let me know if you want to give it a try by yourself or if you need my help to incorporate these changes.

**Comment by D4N** [comment]
@cgilles Do you plan to work on this further?

**Comment by piponazo** [comment]
> @cgilles Do you plan to work on this further?

We are kind of blocked here until the changes I proposed in libheif are merged (It seems that the maintainers of the project have been a bit idle the last weeks):
https://github.com/strukturag/libheif/pull/168

Afterwards, I will need to revisit the conan package I wrote for that.

**Comment by cryptomilk** [comment]
FYI: I've fixed libheif and 1.6.1 works correctly, however HEIF is a patent minefield.

**Comment by clanmills** [comment]
HEIF may be a _patent minefield_ however that doesn't mean that reading the metadata from HEIF involves patents.

I believe the file-format is ISOBMFF (ISO Base Media File Format ISO/IEC 14496-12).  Reading ISOBMFF cannot involve patent infringement.  Decoding the embedded data blocks could involve patents, however metadata blocks (Exif, IPTC, ICC and XMP) are also public standards which do not involve patents.

I'm unclear about the libraries being linked by Exiv2 to undertake reading the file.  Providing support to use another library cannot be a patent infringement by Exiv2.  The dependent library may infringe patents _(for example, to decode compressed data)_, however providing support in Exiv2 to link that library is not a patent infringement.

**Comment by cryptomilk** [comment]
Using libheif is fine, but none of the Linux distributions will package it and thus exiv2 wont have support for HEIF files. Which isn't that bad, as image viewer wont have support either :-)

I'm pushing for AVIF as this is much more likely the next image codec to be widely used. Even Apple joined AOMedia.

**Comment by cgilles** [comment]
Linux Mageia7 : libheif is present de facto...

**Comment by D4N** [comment]
> I'm unclear about the libraries being linked by Exiv2 to undertake reading the file. Providing support to use another library cannot be a patent infringement by Exiv2. The dependent library may infringe patents _(for example, to decode compressed data)_, however providing support in Exiv2 to link that library is not a patent infringement.

I'd love if it were like that, but unfortunately law is not that simple :( I've been told that if you even modify your own library to link against a library that infringes patents, you're putting yourself at risk. So while libheif claims that they don't infringe patents, that hasn't been verified by a lawyer and by supporting linking against it, we risk getting removed from Linux distros that are more restrictive with respect to licensing (Fedora, Debian, openSUSE, etc.)

**Comment by clanmills** [comment]
You're right.  Exiv2 could be booted out the distros and we don't want that to happen.  If we had our own code to read ISOBMFF we would avoid HEIF patent issues.  It's possible we already have code for another format (eg JP2) that only needs a little tweak.

**Comment by boardhead** [comment]
A small addition to an MP4 parser gives you HEIF support

**Comment by cryptomilk** [comment]
A general solution would be better as it can be use with AVIF too. So if you have a MP4 parser already and could extend it, then you can support HEIF and AVIF without linking in more libraries.

**Comment by clanmills** [comment]
When I investigated CR3 (about 2 years ago), I discovered a GitHub repos for isobmffdump by pyke359.  His code is short and to the point.  https://github.com/pyke369/isobmffdump

I invited him to join Team Exiv2 with the following reply:  https://github.com/pyke369/isobmffdump/issues/1#event-1612424215

```
675 rmills@rmillsmbp:~/gnu/github/isobmff/pyke369/isobmffdump $ wc *.cpp
      83     250    2194 dumper.cpp
     260    1006    8054 isobmffdump.cpp
      87     224    2158 tiffs.cpp
     430    1480   12406 total
676 rmills@rmillsmbp:~/gnu/github/isobmff/pyke369/isobmffdump $ 
```
Here it's in action on some HEIF sample files:
```
667 rmills@rmillsmbp:~/gnu/github/isobmff/DigiDNA/ISOBMFF/Example-Files $ ../../../pyke369/isobmffdump/isobmffdump IMG1.HEIC 
@0         | ftyp [24]
@24        | meta [3922]
@3946      | mdat [1462352]
@1466298   | end
668 rmills@rmillsmbp:~/gnu/github/isobmff/DigiDNA/ISOBMFF/Example-Files $ ../../../pyke369/isobmffdump/isobmffdump IMG2.HEIC 
@0         | ftyp [24]
@24        | meta [4350]
@4374      | mdat [848580]
@852954    | end
669 rmills@rmillsmbp:~/gnu/github/isobmff/DigiDNA/ISOBMFF/Example-Files $
```
I believe the Exif, XMP, ICC and IPTC data blocks are stored in the mdat box/atom. 
https://forums.developer.apple.com/thread/89759

The file IMG1.HEIC has Exif metadata (including GPS data) which I can see with Preview.app on the Mac.  The GPS data is Lausanne, Switzerland.  Curious, eh, @piponazo ?

I can see the Exif data with my file dump utility:

```
  0x61f0    25072: .....Exif..MM.*.  ->  f0 00 00 00 06 45 78 69 66 00 00 4d 4d 00 2a 00
  0x6200    25088: ................  ->  00 00 08 00 0b 01 0f 00 02 00 00 00 06 00 00 00
  0x6210    25104: ................  ->  92 01 10 00 02 00 00 00 0e 00 00 00 98 01 12 00
  0x6220    25120: ................  ->  03 00 00 00 01 00 06 00 00 01 1a 00 05 00 00 00
  0x6230    25136: ................  ->  01 00 00 00 a6 01 1b 00 05 00 00 00 01 00 00 00
```

So it's buried in the mdat that starts at 4374 (of length 848580).

I don't think we're far off being able to extract the Exif data and parse it in Exiv2.  I'm stuck in a hotel in Sydney, Australia and it's raining hard outside (60mm today).  Maybe I'll investigate this while I'm stuck.

**Comment by 1div0** [comment]
@clanmills nJoy your trip. Flown by A380?

**Comment by clanmills** [comment]
Singapore Boeing 777 LHR->SIN->SYD.  Will visit Tuan and family in Singapore on the return.  Tuan was a GSoC Student with Exiv2 in 2013.  We may have to isolate ourselves when we get home because of this virus.

**Comment by clanmills** [comment]
@cgilles We will do further work on this PR and will adopt a new approach to HEIF.  #1066

The Exiv2 v0.27.3 is on track to ship on 2020-06-30.  Exiv2 v0.27.3 RC1 will ship on 2020-04-30 and is expected to include support for HEIF.
https://github.com/Exiv2/exiv2/issues/1018#issuecomment-604539346

**Comment by clanmills** [comment]
Topic: Exiv2 and ISOBMFF Support
Time: Jun 14, 2020 13:00 London

Join Zoom Meeting
https://us02web.zoom.us/j/82136730279?pwd=M3hCbll4cWN3ellJd2pCZkxjVEx3Zz09

Meeting ID: 821 3673 0279
Password: 1fDNUV

**Comment by piponazo** [comment]
Ey @clanmills , did you organise that meeting with more people? Honestly I did not follow up the discussion about the topic but I would like to hear more about it. If I am not riding my bike that day I will try to show up in the meeting :wink:

**Comment by clanmills** [comment]
Thanks @piponazo.  The meeting is open to anybody who shows up.  I've put the invitation on every issue involving ISOBMFF and HEIF.

I decided to cancel #1066 in response to criticism from both @D4N and @phako.  The subject has not gone away.  The only feedback I've had about Exiv2 v0.27.3 RCs are dozens of emails every day asking "what is the legal problem with reading ISOBMFF?".

**Pull request closed without merging**
